Social Signup: Add email button options

renancarvalho opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #83872
Fixes https://github.com/Automattic/wp-calypso/issues/83872

## Proposed Changes

* Add email quick links when magic link is sent to users, more info https://github.com/Automattic/wp-calypso/issues/83872

<img width="1117" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/c1c69581-7fd8-4273-8aef-6b6f92143008">

<img width="1109" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/07cddb76-6b6b-407f-9b7f-3b89f23dc0e7">


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
“Before / After” screenshots can also be very helpful when the change is visual.
-->

* Pull and run this branch OR use live link
* Navigate on incognito to `/log-in/link` and use  a “fake email” like 'blabla@fake.com to see all the buttons “shortcuts” available:
 - Gmail
 - AOL
 - yahoo
 - Outlook
 - Apple

If users' email is one of the ones mentioned above, we just show that button, if it is unknown, we show all 5.


C/c @crisbusquets, in the designs I noticed that the bottom part (Stay logged in with the Jetpack Mobile App)of the page is gone, I haven't removed it, should I? 🤔 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

renancarvalho committed:
Add email button options

renancarvalho committed:
Async load icons

renancarvalho committed:
Fix link style and small refactor

renancarvalho committed:
Use translations

renancarvalho committed:
Add logs

heavyweight reviewed (changes requested):
If I enter `fake` in the input and submit I get a blank page and js crash
<img width="1721" alt="Screenshot 2023-11-22 at 18 01 21" src="https://github.com/Automattic/wp-calypso/assets/7000684/8474e895-e05d-412c-87e7-0ca1bfd51b53">

alshakero commented on client/login/magic-login/emailed-login-link-successfully.jsx:
```diff
@@ -45,14 +45,8 @@ class EmailedLoginLinkSuccessfully extends Component {
-							emailAddress,
-						},
-				  } )
+				? translate( "We've sent a login link to " )
```
This "to " looks suspicious.

renancarvalho replied:
I have removed the space at the end :)
The message should be like this according to 5p3iZ99zPYfl6ronoTYEXO-fi-407%3A2074

<img width="478" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/0cb8eb88-70c8-41e2-878e-c542b23421d8">

alshakero replied:
The address should be part of the translated string, you can use {{email}} or %s. Please don't concat strings, it doesn't work in many languages.

renancarvalho replied:
I got it. Fixed.

alshakero commented on client/login/magic-login/emailed-login-link-successfully.jsx:
```diff
@@ -63,24 +57,34 @@ class EmailedLoginLinkSuccessfully extends Component {
-					</a>
+					<p>
+						{ translate(
+							" Didn't get the email? You might want to double check if the email address is associated with your account,"
```
```suggestion
							"Didn't get the email? You might want to double check if the email address is associated with your account,"
```

Is there is a reason for the `,` in the end?

renancarvalho replied:
Yes, this is just about to make a short stop in the sentence. I have removed it from the translation and made it part of the `p` as a string
Design reference :
<img width="362" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/345b490c-cb54-498d-ab35-4b6ae37c88aa">

alshakero replied:
Same issue. We shouldn't concat UI strings.

renancarvalho replied:
Fixed.

alshakero commented on client/login/magic-login/magic-login-email/magic-login-email-icon.tsx:
```diff
@@ -0,0 +1,26 @@
+import React, { useState, useEffect } from 'react';
+
+interface MagicLoginEmailIconProps {
+	icon: string;
+}
+
+export function MagicLoginEmailIcon( { icon }: MagicLoginEmailIconProps ) {
+	const [ IconComponent, setIconComponent ] = useState< React.ReactNode | null >( null );
+
+	useEffect( () => {
+		const getIconComponent = async ( iconName: string ) => {
+			try {
+				// Use dynamic import to load the icons.
+				const module = await import( `calypso/components/social-icons/${ iconName }` );
+				const IconComponent = module.default;
+				setIconComponent( <IconComponent /> );
+			} catch ( error ) {
+				setIconComponent( null );
+			}
+		};
+
+		getIconComponent( icon );
+	}, [ icon ] );
+
+	return <>{ IconComponent }</>;
+}
```
You can use `AsyncLoad` component for this. It does this automatically.

renancarvalho replied:
Ahh this is great! Thank you! 👍

alshakero commented on client/login/magic-login/magic-login-email/magic-login-email-wrapper.tsx:
```diff
@@ -0,0 +1,57 @@
+];
+
+export function MagicLoginEmailWrapper( { emailAddress }: MagicLoginEmailWrapperProps ) {
+	const getEmailDomain = ( email: string ): MagicEmailDomainInfo[] => {
+		const domainMatch = email.match( /@(.+)$/ );
+		const domain = domainMatch ? domainMatch[ 1 ].toLowerCase() : null;
+
+		const filteredDomains = knownDomains.filter( ( e ) => e.domain.toLowerCase() === domain );
+
+		// If no matches, we return the known domains.
+		return filteredDomains.length > 0 ? filteredDomains : knownDomains;
+	};
```
You can export and use [this](https://github.com/Automattic/wp-calypso/blob/implement/login-by-site-url/packages/onboarding/src/utils/domain-suggester.ts#L17) function instead ✨

renancarvalho replied:
Nice! Using it. Thank you.

alshakero commented on client/login/magic-login/magic-login-email/magic-login-email-wrapper.tsx:
```diff
@@ -0,0 +1,57 @@
+		return filteredDomains.length > 0 ? filteredDomains : knownDomains;
+	};
+
+	const logEvent = ( domain: string ) => {
+		recordTracksEvent( 'calypso_magic_login_email_click', { domain } );
+	};
```
This recordTrackEvent won't do anything, it just returns an action. You have to dispatch that action.

renancarvalho replied:
You're right, I imported the wrong one, I should use it from `'@automattic/calypso-analytics';`
I have checked on track vigilante extension and see it going through.

alshakero commented on client/login/magic-login/magic-login-email/magic-login-email-wrapper.tsx:
```diff
@@ -0,0 +1,57 @@
+						onClick={ () => logEvent( item.name ) }
+						target="_blank"
+						href={ item.url }
+						rel="noreferrer"
```
Let's add noopener as well. 

```suggestion
						rel="noreferrer noopener"
```

renancarvalho replied:
I have read more about it. Thank you, added.

alshakero commented on client/login/magic-login/magic-login-email/style.scss:
```diff
@@ -0,0 +1,40 @@
+
```
Weird white space.

renancarvalho replied:
Removed

alshakero commented on client/components/social-icons/gmail.jsx:
```diff
@@ -0,0 +1,62 @@
+		return (
+			<svg
+				className={ classNames( 'social-icons social-icons__gmail', {
+					'social-icons--enabled': ! isDisabled,
+					'social-icons--disabled': !! isDisabled,
```
Strange to have two classes? It should be default vs disabled.

renancarvalho replied:
Yes, I have it copied from the existing components but seems nonsense. 👍

alshakero commented on client/components/social-icons/aol.jsx:
```diff
@@ -0,0 +1,39 @@
+import classNames from 'classnames';
+import PropTypes from 'prop-types';
+import { PureComponent } from 'react';
+
+import './style.scss';
+
+export default class AOLIcon extends PureComponent {
+	static propTypes = {
+		isDisabled: PropTypes.bool,
+	};
+
+	static defaultProps = {
+		isDisabled: false,
+	};
+
+	render() {
+		const { isDisabled, ...rest } = this.props;
+
+		return (
+			<svg
+				className={ classNames( 'social-icons social-icons__aol', {
+					'social-icons--enabled': ! isDisabled,
+					'social-icons--disabled': !! isDisabled,
+				} ) }
+				width="31"
+				height="12"
+				viewBox="0 0 31 12"
+				fill="none"
+				xmlns="http://www.w3.org/2000/svg"
+				{ ...rest }
+			>
+				<path
+					d="M4.70589 0L0 11.6863H3.29411L3.84312 10.0392H7.92156L8.47057 11.6863H11.7647L7.13724 0H4.70589ZM22.7451 0V11.6863H25.3333V0H22.7451ZM16.5883 2.58823C13.8824 2.58823 11.8353 4.6902 11.8353 7.29411C11.8353 10.0392 13.9686 12 16.5883 12C19.2078 12 21.3333 10.0392 21.3333 7.29411C21.3333 4.6902 19.2941 2.58823 16.5883 2.58823ZM5.92159 3.68631L7.13724 7.60787H4.70589L5.92159 3.68631ZM16.5883 5 ...
+					fill="black"
+				/>
+			</svg>
+		);
+	}
+}
```
Surprised these are class components? Let's make them TS function components (GPT can do this).

renancarvalho replied:
Done.

alshakero commented on client/components/social-icons/gmail.jsx:
```diff
@@ -0,0 +1,62 @@
+import classNames from 'classnames';
+import PropTypes from 'prop-types';
+import { PureComponent } from 'react';
+
+import './style.scss';
+
+export default class GmailIcon extends PureComponent {
```
+1.

alshakero commented on client/components/social-icons/outlook.jsx:
```diff
@@ -0,0 +1,138 @@
+import classNames from 'classnames';
+import PropTypes from 'prop-types';
+import { PureComponent } from 'react';
+
+import './style.scss';
+
+export default class OutlookIcon extends PureComponent {
```
+1

alshakero commented on client/components/social-icons/yahoo.jsx:
```diff
@@ -0,0 +1,39 @@
+import classNames from 'classnames';
+import PropTypes from 'prop-types';
+import { PureComponent } from 'react';
+
+import './style.scss';
+
+export default class YahooIcon extends PureComponent {
```
+1

alshakero commented on client/login/magic-login/emailed-login-link-successfully.jsx:
```diff
@@ -63,24 +57,34 @@ class EmailedLoginLinkSuccessfully extends Component {
+							} ) }
+							onClick={ this.onClickBackLink }
+						>
+							{ translate( 'or login with a password instead.' ) }
```
The original version is probably better. Some accounts are passwordless.

renancarvalho replied:
Would you mind elaborating a bit more on what you see as improvements?

alshakero replied:
Not all accounts have passwords, "use password instead" isn't accurate.

renancarvalho replied:
Maybe instead of `or login with a password instead`. And point back to log in, we could say something like or try to reset your password and point the user to https://wordpress.com/wp-login.php?action=lostpassword
WDYT?

renancarvalho replied:
@crisbusquets I would be truly interested in your opinion on this one too :)

crisbusquets replied:
Oh, I see. Instead of assuming they've lost their password (maybe they don't even have a password, as @alshakero mentioned), I'd use the magic link (https://wordpress.com/log-in/link).

What about:

`Didn't get the email? You might want to double check if the email address is associated with your account, or [request a login link] to log you in right away.`

PS: Wow, this interface (https://wordpress.com/wp-login.php?action=lostpassword) is super old!

renancarvalho replied:
@crisbusquets thank you for your input. I think in the context of the message and where users are, requesting a magic link. We should either help users if they did not get an email by sending them to [lostpassword] (https://wordpress.com/wp-login.php?action=lostpassword). Yes, it looks terrible, but it is in the roadmap https://github.com/Automattic/wp-calypso/issues/33924 or/and just point them back to Log-in, which does not help much but make sense IMO.

If I understood correctly, The message you proposed in this context can create an infinite loop since they are already requesting a login link, aka, magic link.

WDYT?

crisbusquets replied:
> @crisbusquets thank you for your input. I think in the context of the message and where users are, requesting a magic link. We should either help users if they did not get an email by sending them to [lostpassword] (https://wordpress.com/wp-login.php?action=lostpassword). Yes, it looks terrible, but it is in the roadmap #33924 or/and just point them back to Log-in, which does not help much but make sense IMO.
> 
> If I understood correctly, The message you proposed in this context can create an infinite loop since they are already requesting a login link, aka, magic link.
> 
> WDYT?

Ahh... yes. For some reason, I didn't see that we were already on the magic link flow, sorry. Too many sign-up and login flows hahahaa

We need to be able to provide a way to log in even if the magic link fails. The lost password flow (what you initially suggested) could work, and it seems the only solution we can offer 🤔 because asking them to use the Jetpack app will overcomplicate everything. Just to know, do we know how many of our users have no password?

renancarvalho replied:
Can we do `Didn't get the email? You might want to double check if the email address is associated with your account, or reset your password` ? Pointing to [this](https://wordpress.com/wp-login.php?action=lostpassword) 
Regarding how many users have no password, I will check that info and get back to you soon.

simison replied:
@crisbusquets @renancarvalho  I think for this convo it’s enough to understand we have a lot of both kinds of users, knowing more won’t really inform the way or the other? :-) 

If we have their email, we know if they're passwordless, and should be able to adjust copy accordingly? They won't be able to create a new password without access to email anyway, and they should contact support.

alshakero commented on client/login/magic-login/magic-login-email/magic-login-email-content.tsx:
```diff
@@ -0,0 +1,14 @@
+	const translate = useTranslate();
+	return (
+		<p>
+			{ translate( 'Open in ' ) } { name }
```
Not good. It should be
```suggestion
			{ translate( 'Open in {{emailProvider/}}') } 
```

This would fail in many languages. In Polish, Open in Gmail would be "Otwórz w Gmailu", notice the "u" after gmail. Translators need to know the context.

renancarvalho replied:
Done

alshakero reviewed (changes requested):
A few issues here and there. Pretty good overall 👍🏼

renancarvalho committed:
Fix translation strings

renancarvalho committed:
Use async load

renancarvalho committed:
Use onboarding package for domains

renancarvalho committed:
Use correct trackEvent import

renancarvalho committed:
Use noopener

renancarvalho committed:
Remove white space from style

renancarvalho committed:
Refactor class to functional components

renancarvalho committed:
Fix translation

renancarvalho commented:
> If I enter `fake` in the input and submit I get a blank page and js crash <img alt="Screenshot 2023-11-22 at 18 01 21" width="1721" src="https://user-images.githubusercontent.com/7000684/284982800-8474e895-e05d-412c-87e7-0ca1bfd51b53.png">

This is a great catch! It is not related to this PR, I have checked prod, and it happens there too. Would you mind if I keep this PR focused and tackle this issue in a following up one?
It must be fixed for sure :)

alshakero commented:
Is it the same issue though? In prod it sends an email to bad emails but it doesn't crash.

renancarvalho commented:
> Is it the same issue though? In prod it sends an email to bad emails but it doesn't crash.

Hmm true, it is crashing… The issue is the same, emails should be validated before reaching into the success email component, which is using this newly added button. Will fix in this PR.

renancarvalho committed:
Fix translations strings

renancarvalho committed:
Fix translations component string

renancarvalho commented:
> Is it the same issue though? In prod it sends an email to bad emails but it doesn't crash.

By using [this function ](https://github.com/Automattic/wp-calypso/blob/implement/login-by-site-url/packages/onboarding/src/utils/domain-suggester.ts?rgh-link-date=2023-11-22T18%3A17%3A33Z#L17), we don't have it crashing anymore, I was not aware that we could send magic links to usernames as well. I have noticed that not all usernames can use magic login, but this is not related to this PR at all. It behaves like before. 👍

crisbusquets commented:
> c/c @crisbusquets, in the designs I noticed that the bottom part (Stay logged in with the Jetpack Mobile App)of the page is gone, I haven't removed it, should I? 🤔

Sorry @renancarvalho I missed this. I'd keep it to be consistent with the magic link flow.

renancarvalho committed:
Change jsx icons component to tsx

renancarvalho committed:
Use to lost password link

renancarvalho committed:
Rename variable

renancarvalho committed:
Fix link line height

renancarvalho committed:
Small style refactor

alshakero reviewed (approved):
Looks great now! I didn't have a chance to test but the code looks great!

crisbusquets commented:
Thanks @renancarvalho !

Looks great :) There are two small things:
- To improve the hierarchy and differentiate between main actions (check the email) and secondary actions (banner) I think we should add a bit more spacing between the `Didn't get the email?...` paragraph and the Jetpack app. In this example I modified the `.card .app-promo .magic-link-app-promo` margin-top to 45px (instead of 4px). The result is something like 48 I think:

<img width="1616" alt="Screenshot 2023-11-24 at 09 28 50" src="https://github.com/Automattic/wp-calypso/assets/25607244/5232257c-f2ae-4812-9d9b-022841d3edbd">

- I'm not 100% convinced about adding all 5 options (Apple, Google, and Outlook seem enough). But if we have tracking, we'll be able to see which ones are most clicked and remove the others.

renancarvalho committed:
Improve element spacing

renancarvalho commented:
> Thanks @renancarvalho !
> 
> Looks great :) There are two small things:
> 
> * To improve the hierarchy and differentiate between main actions (check the email) and secondary actions (banner) I think we should add a bit more spacing between the `Didn't get the email?...` paragraph and the Jetpack app. In this example I modified the `.card .app-promo .magic-link-app-promo` margin-top to 45px (instead of 4px). The result is something like 48 I think:
> 
> <img alt="Screenshot 2023-11-24 at 09 28 50" width="1616" src="https://user-images.githubusercontent.com/25607244/285401091-5232257c-f2ae-4812-9d9b-022841d3edbd.png">
> * I'm not 100% convinced about adding all 5 options (Apple, Google, and Outlook seem enough). But if we have tracking, we'll be able to see which ones are most clicked and remove the others.

The spacing looks better 👍 I have added that.
Regarding the tracking, yes I have added that, soon we can gather this data and iterate over it. 👍

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9739416

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @renancarvalho for including a screenshot in the description! This is really helpful for our translators.

alshakero reviewed (approved):
Tested every case and all looks great!

dlind1 commented on client/login/magic-login/emailed-login-link-successfully.jsx:
```diff
@@ -63,24 +45,34 @@ class EmailedLoginLinkSuccessfully extends Component {
-					</a>
+					<p>
+						{ translate(
+							"Didn't get the email? You might want to double check if the email address is associated with your account,{{a}}or reset your password.{{/a}}",
```
```suggestion
							"Didn't get the email? You might want to double check if the email address is associated with your account, {{a}}or reset your password.{{/a}}",
```
I think that there should be a space after the comma?

renancarvalho replied:
Hi 👋 
The spacing is done using https://github.com/Automattic/wp-calypso/pull/84431/files#diff-f3418ecf5bbbe75b9b54d30784a467869f49ef4dd83f80c80c778809117d3193R193.

<img width="439" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/5d1af47f-f82f-4cd4-bd0a-eaff0ab5f7d3">

dlind1 replied:
It's a small issue since in a lot of translations there is no comma or the link is in a different place, but some translators might assume a typo and add a space after a comma anyway. Just for future consideration 🙂

renancarvalho committed:
Merge branch 'trunk' into feat/quick-email

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9739416).
